proposals/0000-Move-Platforms-to-Own-Packages.md

```diff
@@ -0,0 +1,273 @@
+   * `Systrace`
+   * `TVEventHandler`
+   * `UIManager`
+   * `unstable_batchedUpdates`
+   * `Vibration`
+   * `YellowBox`
+   * **Plugins**
+   * `DeviceEventEmitter`
+   * `NativeAppEventEmitter`
+   * `NativeModules`
+   * `Platform`
+   * `processColor`
+   * `requireNativeComponent`
+   * `takeSnapshot`
+   * **Deprecated PropTypes**
+   * `ColorPropType`
+   * `EdgeInsetsPropType`
+   * `PointPropType`
+   * `ViewPropTypes`
+
+ * `react-native-ios`
+   * **Components**
+   * `DatePickerIOS`
+   * `InputAccessoryView`
+   * `MaskedViewIOS`
+   * `PickerIOS`
+   * `ProgressViewIOS`
+   * `SegmentedControlIOS`
+   * `SnapshotViewIOS`
+   * `TabBarIOS`
+   * **APIs**
+   * `ActionSheetIOS`
+   * `AlertIOS`
+   * `ImagePickerIOS`
+   * `PushNotificationIOS`
+   * `StatusBarIOS`
+   * `VibrationIOS`
+
+ * `react-native-android`
+   * **Components**
+   * `DrawerLayoutAndroid`
+   * `ProgressBarAndroid`
+   * `ToastAndroid`
+   * `ToolbarAndroid`
+   * `ViewPagerAndroid`
+   * **APIs**
+   * `DatePickerAndroid`
+   * `PermissionsAndroid`
+   * `TimePickerAndroid`
+
+
+### Hypothetical project overview (non-exhaustive)
+
+This is the *very basic* overview of what the GitHub repo could look like:
+
+```
+-| project root
+--📁 {.circleci, .github, bots}/
+--📁 flow/
+--📁 flow-github/
+--📁 packages/
+---📁 react-native/
+----📁 Libraries/ (All cross-platform JS)
+----📁 ReactCommon/
+----📁 scripts/
+-----📄 launchPackager.bat
+-----📄 launchPackager.command
+-----📄 packager.sh
+----📁 third-party/
+----📁 tools/ (contains bzl build-defs)
+---📁 react-native-ios/
+----📁 Libraries/
+----📁 React/
+----📁 scripts/
+-----📄 ios-configure-glog.sh
+-----📄 ios-install-third-party.sh
+-----📄 react-native-xcode.sh
+----📁 third-party-podspecs/
+---📁 react-native-android/
+----📁 gradle/
+----📁 keystores/
+----📁 Libraries/
+----📁 ReactAndroid/
+---📁 react-native-git-upgrade/
+--📁 RNTester/
+--📁 scripts/ (CI, sync, etc. scripts)
+--📄 package.json (for the Yarn workspace)
+--📄 {readme, contributing, CoC, other markdown docs}
+```
+
+I am not fully sure what the layout of the local-cli stuff will end up needing to be after this; with [this proposal](https://github.com/react-native-community/discussions-and-proposals/blob/master/proposals/0002-Extracted-Command-Line-Tools.md) the bulk of it will end up getting moved to another repo anyways.
+
+ReactIOS and ReactAndroid would get entries in their `package.json` like this:
+
+```json
+{
+  "rnpm": {
+    "haste": {
+      "providesModuleNodeModules": ["react-native-ios"],
+      "platforms": ["ios"]
+    }
+  }
+}
+```
+
+Platform-specific Flow prop types will probably need to stay bundled together in the main package. Ideally it could be set up so if you didn't have `react-native-ios` installed the iOS-only View props wouldn't show up when typechecking, for example. But I don't know if there's a way to do that with Flow yet. Would need [this issue](https://github.com/facebook/flow/issues/396) to be resolved, probably.
```

Comment:
How do other platforms like windows extend the props of components to provide platform specific props?

Reply:
In the case of RNWindows they use the platform-overridable `prop-types` files:

https://github.com/Microsoft/react-native-windows/blob/master/Libraries/Components/View/PlatformViewPropTypes.windows.js
https://github.com/Microsoft/react-native-windows/blob/master/Libraries/Text/TextPropTypes.windows.js

They don't seem to use platform overridable flow props anywhere. But a platform-specific `X-Types.platform.js` would be awkward, because you wouldn't be able to see the flow types for `windows` when flow checking `*.ios.js` files for example.